esp-hal/src/soc/esp32/psram.rs

```diff
@@ -9,53 +9,98 @@
 //! present on the `ESP32` chip. `PSRAM` provides additional external memory to
 //! supplement the internal memory of the `ESP32`, allowing for increased
 //! storage capacity and improved performance in certain applications.
-//!
-//! The `PSRAM` module is accessed through a virtual address, defined as
-//! `PSRAM_VADDR`. The starting virtual address for the PSRAM module is
-//! 0x3F800000. The `PSRAM` module size depends on the configuration specified
-//! during the compilation process. The available `PSRAM` sizes are `2MB`,
-//! `4MB`, and `8MB`.
-//!
-//! NOTE: If you want to use `PSRAM` on `ESP32` or `ESP32-S3`, it'll work only
-//! in `release` mode.
 
-/// The starting virtual address of the PSRAM (Pseudo-SRAM) region in memory.
-pub const PSRAM_VADDR_START: usize = 0x3F800000;
+pub use crate::soc::psram_common::*;
+
+const EXTMEM_ORIGIN: usize = 0x3F800000;
+
+/// Cache Speed
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Default)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[allow(missing_docs)]
+pub enum PsramCacheSpeed {
+    #[default]
+    PsramCacheF80mS40m = 0,
+    PsramCacheF40mS40m,
+    PsramCacheF80mS80m,
+}
 
-/// Retrieves the starting virtual address of the PSRAM (Pseudo-SRAM) region in
-/// memory.
-pub fn psram_vaddr_start() -> usize {
-    PSRAM_VADDR_START
+/// PSRAM configuration
+#[derive(Copy, Clone, Debug, Default)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct PsramConfig {
+    /// PSRAM size
+    pub size: PsramSize,
+    /// Cache speed
+    pub cache_speed: PsramCacheSpeed,
 }
 
-cfg_if::cfg_if! {
-    if #[cfg(feature = "psram-2m")] {
-        const PSRAM_SIZE: u32 = 2;
-    } else if #[cfg(feature = "psram-4m")] {
-        const PSRAM_SIZE: u32 = 4;
-    } else if #[cfg(feature = "psram-8m")] {
-        const PSRAM_SIZE: u32 = 8;
+/// Initializes the PSRAM memory on supported devices.
+///
+/// Returns the start of the mapped memory and the size
+pub fn init_psram(_peripheral: crate::peripherals::PSRAM, config: PsramConfig) -> (*mut u8, usize) {
+    let mut config = config;
+
+    utils::psram_init(&config);
+
+    if config.size.is_auto() {
+        // Reading the device-id turned out to not work as expected (some bits flipped
+        // for unknown reason)
+        //
+        // As a workaround we just map 4m (maximum we can do) and
+        // probe if we can access top of PSRAM - if not we assume it's 2m
+        //
+        // This currently doesn't work as expected because of https://github.com/esp-rs/esp-hal/issues/2182
+        utils::s_mapping(EXTMEM_ORIGIN as u32, 4 * 1024 * 1024);
+
+        let guessed_size = unsafe {
+            let ptr = (EXTMEM_ORIGIN + 4 * 1024 * 1024 - 36 * 1024) as *mut u8;
+            for i in 0..(36 * 1024) {
```

Comment:
Why 36K?

Reply:
Idea is to write and read more than fits the cache

Ideally we get reading the chip id to work - that would be the best solution